bfdd/bfdd_northbound.c

```diff
@@ -0,0 +1,1118 @@
+	return bfd_session_destroy(event, dnode, false);
+}
+
+static const void *
+bfdd_bfd_sessions_single_hop_get_next(const void *parent_list_entry
+				      __attribute__((__unused__)),
+				      const void *list_entry)
+{
+	return bfd_session_next(list_entry, false);
+}
+
+static int bfdd_bfd_sessions_single_hop_get_keys(const void *list_entry,
+						 struct yang_list_keys *keys)
+{
+	const struct bfd_session *bs = list_entry;
+	char dstbuf[INET6_ADDRSTRLEN];
+
+	inet_ntop(bs->key.family, &bs->key.peer, dstbuf, sizeof(dstbuf));
+
+	keys->num = 3;
+	strlcpy(keys->key[0], dstbuf, sizeof(keys->key[0]));
+	strlcpy(keys->key[1], bs->key.ifname, sizeof(keys->key[1]));
+	strlcpy(keys->key[2], bs->key.vrfname, sizeof(keys->key[2]));
+
+	return NB_OK;
+}
+
+static const void *
+bfdd_bfd_sessions_single_hop_lookup_entry(const void *parent_list_entry
+					  __attribute__((__unused__)),
+					  const struct yang_list_keys *keys)
+{
+	const char *dest_addr = keys->key[0];
+	const char *ifname = keys->key[1];
+	const char *vrf = keys->key[2];
+	struct sockaddr_any psa, lsa;
+	struct bfd_key bk;
+
+	strtosa(dest_addr, &psa);
+	memset(&lsa, 0, sizeof(lsa));
+	gen_bfd_key(&bk, &psa, &lsa, false, ifname, vrf);
+
+	return bfd_key_lookup(bk);
+}
+
+/*
+ * XPath: /frr-bfdd:bfdd/bfd/sessions/single-hop/source-addr
+ */
+static int bfdd_bfd_sessions_single_hop_source_addr_modify(
+	enum nb_event event __attribute__((__unused__)),
+	const struct lyd_node *dnode __attribute__((__unused__)),
+	union nb_resource *resource __attribute__((__unused__)))
+{
+	return NB_OK;
+}
+
+static int bfdd_bfd_sessions_single_hop_source_addr_destroy(
+	enum nb_event event __attribute__((__unused__)),
+	const struct lyd_node *dnode __attribute__((__unused__)))
+{
+	return NB_OK;
+}
+
+/*
+ * XPath: /frr-bfdd:bfdd/bfd/sessions/single-hop/detection-multiplier
+ */
+static int bfdd_bfd_sessions_single_hop_detection_multiplier_modify(
+	enum nb_event event, const struct lyd_node *dnode,
+	union nb_resource *resource __attribute__((__unused__)))
+{
+	uint8_t detection_multiplier = yang_dnode_get_uint8(dnode, NULL);
+	struct bfd_session *bs;
+
+	switch (event) {
+	case NB_EV_VALIDATE:
+		if (detection_multiplier == 1)
+			return NB_ERR_VALIDATION;
+		break;
+
+	case NB_EV_PREPARE:
+		/* NOTHING */
+		break;
+
+	case NB_EV_APPLY:
+		bs = nb_running_get_entry(dnode, NULL, true);
+		bs->detect_mult = detection_multiplier;
+		break;
+
+	case NB_EV_ABORT:
+		/* NOTHING */
+		break;
+	}
+
+	return NB_OK;
+}
+
+/*
+ * XPath: /frr-bfdd:bfdd/bfd/sessions/single-hop/desired-transmission-interval
+ */
+static int bfdd_bfd_sessions_single_hop_desired_transmission_interval_modify(
+	enum nb_event event, const struct lyd_node *dnode,
+	union nb_resource *resource __attribute__((__unused__)))
+{
+	uint32_t tx_interval = yang_dnode_get_uint32(dnode, NULL);
+	struct bfd_session *bs;
+
+	switch (event) {
+	case NB_EV_VALIDATE:
+		if (tx_interval < 10000 || tx_interval > 60000000)
+			return NB_ERR_VALIDATION;
```

Comment:
Similarly, you could add a YANG "range" statement to the "desired-transmission-interval" leaf to perform these validations there. Code-level validations should be used only when strictly necessary.

Reply:
This is a FRR software limitation: BFD specification and YANG model RFC tells to use microseconds, but we can't do that with current software. So instead of limiting the protocol in specification, we are limiting on the software. It is possible that a day comes and someone implements BFD microseconds timer support (with hardware or real-time OS probably).

Reply:
Makes sense to me. Will mark this as resolved...